command/update/command.go

```diff
@@ -269,8 +201,48 @@ func (c *Command) execute() error {
 			return microerror.MaskAny(err)
 		}
 
-		c.logger.Log("debug", fmt.Sprintf("updated endpoint for service '%s'", f.Kubernetes.Cluster.Service))
+		c.logger.Log("debug", fmt.Sprintf("added IP to endpoint of service '%s'", f.Kubernetes.Cluster.Service))
 	}
 
+	// Listen to OS signals issued by the Kubernetes scheduler.
+	listener := make(chan os.Signal, 2)
+	signal.Notify(listener, syscall.SIGTERM, syscall.SIGKILL)
+
+	c.logger.Log("debug", "waiting for termination signal")
+
+	s := <-listener
+
+	c.logger.Log("debug", fmt.Sprintf("received termination signal: %#v (%s)\n", s, s))
+
+	// Use the updater to actually delete the endpoints identified by the provided
+	// flags.
+	go func() {
+		action := func() error {
+			err := newUpdater.Delete(f.Kubernetes.Cluster.Namespace, f.Kubernetes.Cluster.Service, podInfos)
+			if err != nil {
+				return microerror.MaskAny(err)
+			}
+
+			return nil
+		}
+
+		err := backoff.Retry(action, backoff.NewExponentialBackOff())
+		if err != nil {
+			c.logger.Log("error", fmt.Sprintf("%#v", microerror.MaskAny(err)))
+			os.Exit(1)
+		}
+
+		c.logger.Log("debug", fmt.Sprintf("removed IP from endpoint of service '%s'", f.Kubernetes.Cluster.Service))
+		c.logger.Log("debug", "shutting down with exit code 0")
+
+		os.Exit(0)
+	}()
+
+	<-listener
```

Comment:
I saw folks are using `for` in long running receives. I don't know exact reason, but I'd also feel safer with that. Maybe there is some timeout in receive or it can be interrupted because of kernel magic.

Reply:
That was just copied from microkit. See https://github.com/giantswarm/microkit/blob/4871f763488a16c5b7ea3c411bbd82cd5139e29a/command/daemon/command.go#L137-L156.

Reply:
So we are using this every day in all of our running microservices. The signal notifier is a go primitive so I am not sure what should go wrong there. See https://golang.org/pkg/os/signal/#Notify.

Reply:
I won't argue here. I don't have enough knowledge.